xkcd/cmd/main.go

```diff
@@ -0,0 +1,42 @@
+// xkcd is an xkcd offline indexing and quering tool.
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/vespian/go-exercises/xkcd/pkg/cmdline"
+	"github.com/vespian/go-exercises/xkcd/pkg/index"
+	"github.com/vespian/go-exercises/xkcd/pkg/types"
+)
+
+func main() {
+	var err error
+
+	c := cmdline.Parse()
+	fmt.Printf("Commandline: %+v", c)
+
+	switch c.Op {
+	case types.Update:
+		err = index.Update(c.XkcdURI, c.Range, c.IndexFile)
+	case types.List:
+		var d types.AllStories
+		if d, err = index.Fetch("", c.Range, c.IndexFile); err == nil {
+			fmt.Print(d)
+		}
+	case types.Search:
+		var d types.AllStories
+		if d, err = index.Fetch(c.QueryString, c.Range, c.IndexFile); err == nil {
+			fmt.Println(d)
+		}
+	default:
+		// Should not happen, but still, just in case:
+		fmt.Fprintf(os.Stderr, "Unsuporrted operation: `%s`\n", c.Op)
+		os.Exit(1)
+	}
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Operation failed: %s\n", err)
+		os.Exit(1)
```

Comment:
Note that os.Exit does *not* run defer'd functions. So in this case it's safe, but a better approach is to have a top-level function that runs something like Main() (note the capital letter), a function which returns an error, and then os.Exit()'s in that case.

Say something like this.

```
func main() {
  if err := doMain(); err != nil {
    os.Exit(1)
  }
}

func doMain() error {
  defer someCleanup()
  ...
  defer someMoreCleanup()
  ...
  if err != nil {
    return err
  }
  return nil
}
```